src/cryptography/hazmat/bindings/openssl/binding.py

```diff
@@ -172,3 +172,11 @@ def _lock_cb(cls, mode, n, file, line):
                     mode, n, file, line
                 )
             )
+
+
+# OpenSSL is not thread safe until the locks are initialized. We call this
+# method in module scope so that it executes with the import lock. On
+# Pythons < 3.4 this import lock is a global lock, which can prevent a race
+# condition registering the OpenSSL locks. On Python 3.4+ the import lock
+# is per module so this approach will not work.
```

Comment:
It definitely doesn't fix it, but it may help some users. I'm still completely at a loss for how to fix this without just doing it all at the C layer.